### PR TITLE
[docs] Clarification about indexed event parameters

### PR DESCRIPTION
**Pull request by cameel**

### CODE REVIEW

File: docs/contracts/events.rst [hunk 1]
```diff
@@ -61,7 +62,12 @@ The hash of the signature of the event is one of the topics, except if you
 declared the event with the ``anonymous`` specifier. This means that it is
 not possible to filter for specific anonymous events by name, you can
 only filter by the contract address. The advantage of anonymous events
-is that they are cheaper to deploy and call.
+is that they are cheaper to deploy and call. It also allows you to declare
+four indexed arguments rather than three.
+
+.. note::
+    In the transaction log a non-anonymous event is indistinguishable from an anonymous one that
```
Comment (chriseth) [inline docs/contracts/events.rst]: Maybe rephrase to something like `Since the transaction log only stores the event data and not the type, you have to know the type of the event, including which parameter is indexed and if the event is anonymous or not in order to correctly interpret the data. In particular, it is possible to "fake" the signature of another event using an anonymous event.`

Comment (cameel) [reply docs/contracts/events.rst]: I actually wanted to include the part about faking an event in my version initially but in the end I decided to keep it short :)

Done.

**Comment by cameel** [comment]
@aarlt has pointed out that the [docs on ABI encoding of events](https://docs.soliditylang.org/en/v0.8.4/abi-spec.html#events) say

> If the event is declared as anonymous the `topics[0]` is not generated;

This could be taken to mean that `topic[0]` is left empty in that case. I think it just does not contain the signature but it can still contain one of the indexed parameters. Not sure if the first one or the fourth one when it's there (both would make sense). I need to check that and clarify that part too.

**Comment by aarlt** [comment]
Yes, I did not understand the section correctly. When I read in https://docs.soliditylang.org/en/v0.8.4/abi-spec.html#events

- topics[0]: keccak(EVENT_NAME+"("+EVENT_ARGS.map(canonical_type_of).join(",")+")") (canonical_type_of is a function that simply returns the canonical type of a given argument, e.g. for uint indexed foo, it would return uint256). If the event is declared as anonymous the topics[0] is not generated;
- topics[n]: abi_encode(EVENT_INDEXED_ARGS[n - 1]) (EVENT_INDEXED_ARGS is the series of EVENT_ARGS that are indexed);

I understood it like `topics[0]` is only filled, when a non-anonymous event is used. And that the indexed topics are stored from `topics[1]` on - because if I use the specification `topics[n]: abi_encode(EVENT_INDEXED_ARGS[n - 1]) ` would be for `topic[0]`, `abi_encode(EVENT_INDEXED_ARGS[-1])` that don't made sense for me. So I thought that it is only valid for `n>0`. e.g. `topics[1] = EVENT_INDEXED_ARGS[0]`.

@cameel Thanks for your awesome reviews!

File: docs/contracts/events.rst [hunk 2]
```diff
@@ -61,7 +62,16 @@ The hash of the signature of the event is one of the topics, except if you
...
 not possible to filter for specific anonymous events by name, you can
 only filter by the contract address. The advantage of anonymous events
-is that they are cheaper to deploy and call.
+is that they are cheaper to deploy and call. It also allows you to declare
+four indexed arguments rather than three.
+
+.. note::
+    Since the transaction log only stores the event data and not the type,
+    you have to know the type of the event, including which parameter is
+    indexed and if the event is anonymous or not in order to correctly
```
Comment (hrkrshnn) [inline docs/contracts/events.rst]:
```suggestion
    indexed and if the event is anonymous in order to correctly
```

File: docs/contracts/events.rst [hunk 3]
```diff
@@ -24,9 +24,10 @@ because the contract can only see the last 256 block hashes.
 
 You can add the attribute ``indexed`` to up to three parameters which adds them
 to a special data structure known as :ref:`"topics" <abi_events>` instead of
-the data part of the log. If you use arrays (including ``string`` and ``bytes``)
-as indexed arguments, its Keccak-256 hash is stored as a topic instead, this is
-because a topic can only hold a single word (32 bytes).
+the data part of the log.
+A topic can only hold a single word (32 bytes) so if you use a :ref:`reference type
+<reference-types>` for an indexed argument, Keccak-256 hash of the value is stored
```
Comment (hrkrshnn) [inline docs/contracts/events.rst]:
```suggestion
<reference-types>` for an indexed argument, the Keccak-256 hash of the value is stored
```

Comment (hrkrshnn) [inline docs/contracts/events.rst]: Perhaps we should list what reference types are supported? I think it would just be `arrays` and `structs`.

Comment (cameel) [reply docs/contracts/events.rst]: One of the reasons for this PR is that I noticed that structs are not mentioned and I replaced the text to avoid having the same problem in the future when we add another reference type :)

But I can add that back if you think it was better.

Comment (hrkrshnn) [reply docs/contracts/events.rst]: Okay, but we don't support all of them currently, right? For example `mapping`. 

Yeah, I'm fine with both. The user would get an error when they use `mapping` or other reference type anyway.

Comment (chriseth) [reply docs/contracts/events.rst]: Yeah, I think "reference type" is fine without listing all of them.

Comment (cameel) [reply docs/contracts/events.rst]: > Okay, but we don't support all of them currently, right? For example mapping.

I think we do. It's surprising but mapping is not actually a reference type. It's not listed under "Reference Types" in the docs and also in the code `MappingType` does not inherit from `ReferenceType`.

File: docs/contracts/events.rst (hunk 2 above)
Comment (cameel) [reply docs/contracts/events.rst]: Fixed.

File: docs/contracts/events.rst (hunk 3 above)
Comment (cameel) [reply docs/contracts/events.rst]: Fixed.

**Comment by cameel** [comment]
This is ready for another review.

I checked how the topic slots are filled for anonymous events and based on [this part in the IR codegen](https://github.com/ethereum/solidity/blob/aec20f0038e6a0053855a702e1d2c47dcc0bf2b3/libsolidity/codegen/ir/IRGeneratorForStatements.cpp#L980-L1018) I see that the first topic is inserted at position 0 (rather than 1), the second topic at position 1 (rather than 2) and so on. I have adjusted the ABI docs to be clearer about it.

**Review by hrkrshnn — APPROVED** [review]
